source/exe/win32/platform_impl.cc

```diff
@@ -1,17 +1,60 @@
+#include <chrono>
+#include <thread>
+
+#include "common/buffer/buffer_impl.h"
 #include "common/common/assert.h"
 #include "common/common/thread_impl.h"
+#include "common/event/signal_impl.h"
 #include "common/filesystem/filesystem_impl.h"
 
 #include "exe/platform_impl.h"
 
 namespace Envoy {
 
+static volatile bool shutdown_pending = false;
+
+BOOL WINAPI CtrlHandler(DWORD fdwCtrlType) {
+  if (shutdown_pending) {
+    return 0;
+  }
+  shutdown_pending = true;
+
+  auto handler = Event::eventBridgeHandlersSingleton::get()[ENVOY_WIN32_SIGTERM];
+  if (!handler) {
+    return 0;
+  }
+
+  // This code is executed as part of a thread running under a Windows
+  // context. For that reason we want to avoid allocating memory or
+  // taking locks. This is why we use do not want to just write to a socket
+  // to wake up the signal handler.
+  Buffer::OwnedImpl buffer;
+  constexpr absl::string_view data{"a"};
+  buffer.add(data);
```

Comment:
This allocates memory, contrary to the above comment. I think you can add an UnownedSlice to the buffer, and put the slice on the stack.

Reply:
sure will do

Reply:
I dont see an overload of `add` that takes an `UnownedSlice` as input. There is `addBufferFragment` but that allocates the slice.

Should I just add an overload or a function `addUnownedSlice`?

Reply:
Honestly I am not sure if this 1 byte allocation is going to save us if we are that close to OOM

Reply:
I missed this issue.  The allocation of a buffer object ends up allocating memory too.  I would suggest use of writev instead of write from a buffer.

My understanding of the no-allocation rule in signal handlers is somewhat crude.  I think that the issue is that the allocation process itself takes locks; if the signal handler triggers while the thread is holding the allocator mutex the signal handler deadlocks.

Reply:
I changed it into `writev` and I opened another issue for some other windows threads that call `write` instead of `writev`